### ISSUE TITLE
clang-tidy: bugprone-unhandled-self-assignment false positive

### ISSUE THREAD

**Issue by kalinnikovv** [llvm/llvm-project#53313]
New bugprone-unhandled-self-assignment check provides false positives in some cases. It seems to check only for the presence of specific constructs which makes it less useful.

Example code with explanation:
```c++
template<typename T>
struct MyPointer
{
    T* p;
    
    MyPointer& operator=(const MyPointer& other) // warning: operator=() does not handle self-assignment properly [bugprone-unhandled-self-assignment]
    {
        if (p != other.p) // however, on self-assignment p will be equal to other.p so this block will be skipped
        {
            ref(other.p);
            T* old = p;
            p  = other.p;
            deref(p);
        }
        return *this;
    }
    
    static void ref(T* ptr)
    {
        ++ptr->refCtr;
    }
    
    static void deref(T* ptr)
    {
        if (!--ptr->refCtr)
            delete ptr;
    }
};
```

**Comment by kalinnikovv** [comment]
More complete code example:

```c++
struct S {
    std::atomic_int refCtr;
};

template<typename T>
struct IntrusivePointer
{
    T* p = nullptr;

    IntrusivePointer(const IntrusivePointer& other)
        : p(other.P)
    {
        ref(p);
    }

    ~IntrusivePointer()
    {
        deref(p);
    }

    IntrusivePointer& operator=(const IntrusivePointer& other)
    {
        if (p != other.p)
        {
            T* old = p;
            p = other.p;
            ref(p);
            deref(old);
        }
        return *this;
    }

    void swap(IntrusivePointer& other)
    {
         std::swap(p, other.p);
    }

    static ref(T* p)
    {
        if (p)
        {
           ++p->refCtr;
        }
    }

    static deref(T* p)
    {
        if (p && !--p->refCtr)
        {
            delete p;
        }
    }
};
```

**Comment by benui-dev** [comment]
Can confirm this still occurs

**Comment by N-Dekker** [comment]
Using LLVM 15.04 on Windows, false positive unhandled-self-assignment messages appeared when I adjusted the self-assignment check example from https://clang.llvm.org/extra/clang-tidy/checks/bugprone/unhandled-self-assignment.html by adding one more line of code, making it a class _template_ instead:

```
template <typename MyTemplateParameter> // <== My only adjustment to the example (Niels Dekker, December 2022)
class T {
int* p;

public:
  T(const T &rhs) : p(rhs.p ? new int(*rhs.p) : nullptr) {}
  ~T() { delete p; }

  // ...

  T& operator=(const T &rhs) {
    if(this == &rhs)
      return *this;

    delete p;
    p = new int(*rhs.p);
    return *this;
  }
};

```

Clang-Tidy says:
> operator=() does not handle self-assignment properly [bugprone-unhandled-self-assignment,cert-oop54-cpp]

So I guess this specific warning does not yet correctly support class templates, right?

**Comment by Mike4Online** [comment]
I am also seeing `bugprone-unhandled-self-assignment` false-positives from Clang Tidy when run against a template class.  Here's the beginning of my copy assignment operator for a templated class named `Vector`:

```
	Vector<T>& operator=(const Vector<T>& other) // Copy assignment operator
	{
		// Guard self assignment
		if (this == &other)
			return *this; // self-assignment

		if (_Myfirst != nullptr)
			delete[] _Myfirst;

		m_max_size = other.m_max_size;
		_Myfirst = new T[m_max_size];

		m_arraySize = other.m_arraySize;
		for (int32_t i = 0; i < m_arraySize; ++i)
			_Myfirst[i] = other[i];

		return*this;
	}
```
But I still get a false-positive on the line where this copy assignment operator definition begins.  This class is implemented entirely within a header file, as is the norm for a templated class, and there is no separate declaration.  Perhaps Clang Tidy is not instantiating the class?  Or expecting a declaration distinct from the definition?

I am running `clang-tidy.exe` version **15.0.7**, as provided by Clang Power Tools and LLVM 15.0.7

**Comment by Mike4Online** [comment]
An update to my previous post, above:

My custom `Vector` class lacks a definition for `operator==`.  This may be why Clang Tidy reports my copy assignment operator does not handle _assignment to self_ properly, even though the `if (this == &other)` test is present.  I assume the compiler provides a default `operator==` for my custom class, and that Clang Tidy somehow knows this equality operator is not sufficient to guard against self-assignment.  But I am not sure yet if this is the underlying issue.

If my premise is correct then there is no bug with `bugprone-unhandled-self-assignment` in my case; however, there may be room for improving the warning text associated with `bugprone-unhandled-self-assignment` when scenarios like mine are the underlying issue.  E.G. suggested warning text for `bugprone-unhandled-self-assignment`:


`operator=() does not handle self-assignment properly.  Check the copy assignment operator for a guard, and operator==() for a thorough compare.  [bugprone-unhandled-self-assignment]`

**Comment by Mike4Online** [comment]
A note about my previous post, above.  My custom `Vector` class currently lacks an equality operator.  That should not matter since the guard within the copy assignment operator is comparing two pointers / addresses to each other:

`if (this == &other)`

At some point I'll add an equality operator to my class and see if that makes any difference for `bugprone-unhandled-self-assignment`.

**Comment by N-Dekker** [comment]
> > Using LLVM 15.04 on Windows, false positive unhandled-self-assignment messages appeared when I adjusted the self-assignment check example from https://clang.llvm.org/extra/clang-tidy/checks/bugprone/unhandled-self-assignment.html by adding one more line of code, making it a class _template_ instead:
...

> This example is no longer reproductible on trunk https://godbolt.org/z/hd9oxxGe5 Original issue reproductible as is.

Thanks for trying it out, @vbvictor 

I can still reproduce the false positive message on [my template example](https://github.com/llvm/llvm-project/issues/53313#issuecomment-1365510724) locally on my Windows laptop, _even_ when using LLVM21.1.6. But only when having `-std=c++14` or `-std=c++17` (not with `-std=c++20`). I don't know why it does not appear on godbolt.org 🤷 

As you say it is _no longer_ reproductible on trunk, did you see it on godbolt.org before?
